app/models/exercise.rb

```diff
@@ -95,6 +95,36 @@ class Exercise < ApplicationRecord
   scope :can_release_to_a15k, -> { where(release_to_a15k: true) }
   scope :not_released_to_a15k, -> { where(a15k_identifier: nil) }
 
+  before_validation :set_context
+
+  def set_context(archive_version: nil)
+    return unless context.nil?
+
+    tag_names = tags.map(&:name)
+    cnxfeature_tags = tag_names.filter { |name| name.starts_with? 'context-cnxfeature:' }
+    cnxmod_tags = tag_names.filter     { |name| name.starts_with? 'context-cnxmod:' }
+    return if cnxfeature_tags.empty? || cnxmod_tags.empty?
+
+    s3 = OpenStax::Content::S3.new
+    archive_version ||= s3.ls.last
+    archive = OpenStax::Content::Archive.new archive_version
+    cnxmod_tags.each do |cnxmod_tag|
+      page_uuid = cnxmod_tag.sub 'context-cnxmod:', ''
+      page = s3.find_page page_uuid, archive_version: archive_version
+      next if page.nil?
+
+      node = Nokogiri::HTML.parse archive.json(page)['content']
+      cnxfeature_tags.each do |cnxfeature_tag|
+        feature_id = cnxfeature_tag.sub 'context-cnxfeature:', ''
+        feature = node.at_css "##{feature_id}"
+        next if feature.nil?
+
+        self.context = feature.to_html
+        return
+      end
+    end
+  end
```

Comment:
This looks good, but could there be a performance concern? I can do some testing.

Reply:
There definitely is an impact of a couple of seconds but it runs when saving so the search should still be fine...
I'll have to run a loop manually once this is deployed to set the context for existing exercises, since the server currently doesn't have background jobs setup

Reply:
I think this looks great, but...   I'm wondering how locked in a para with ID is once it's authored.  Does the wording every change across versions? If so maybe we need to store the version that the content was downloaded as so we can display a little "from content version X" line below the context.  I'll ask content and we may have to make an update to add that.  This is great for v1 though

Reply:
I think currently if you change the feature content you'll need to create a new exercise version to see the change but any import into Tutor will also see the change.

We can easily make Tutor get the context from Exercises instead so what you see in Exercises is what you get in Tutor. We probably need to wait at least 1 version though so I have time to populate the context for existing exercises.